pkg/apis/core/validation/validation.go

```diff
@@ -3008,12 +3008,19 @@ func ValidateNodeSelectorRequirement(rq core.NodeSelectorRequirement, fldPath *f
 func ValidateNodeSelectorTerm(term core.NodeSelectorTerm, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if len(term.MatchExpressions) == 0 {
-		return append(allErrs, field.Required(fldPath.Child("matchExpressions"), "must have at least one node selector requirement"))
+	if len(term.MatchExpressions) == 0 && len(term.MatchFields) == 0 {
```

Comment:
I think this loosening of validation makes sense, but can cause grief for a 1.10-level apiserver (during rolling upgrade or downgrade from running 1.11):

1. an object with a podspec is submitted with MatchFields set and MatchExpressions unset, 1.11 apiserver validates and persists it
2. during rolling HA upgrade or after downgrade to 1.10, a 1.10 apiserver will not know about the MatchFields field, and would consider the object invalid and disallow updates to it (and possibly disallow deleting it)

The doc on the `NodeSelectorTerm` type says `// A null or empty node selector term matches no objects.` which would be acceptable for the 1.10 scheduler to do (would fail safe with "unscheduleable" error), but the actual API validation code in 1.10 prevents an empty node selector term from being allowed.

Reply:
> The doc on the NodeSelectorTerm type says `// A null or empty node selector term matches no objects`. which would be acceptable for the 1.10 scheduler to do (would fail safe with "unscheduleable" error), but the actual API validation code in 1.10 prevents an empty node selector term from being allowed.

Good point; the implementation of schedule match this doc/comment. Scheduler will let pod pending with 'unscheduleable' error. I think we should remove this validation in 1.11; for 1.10, not sure, cherry-pick??

Reply:
> I'm a little confused about this; another meaning maybe: the NodeSelectorTerms []NodeSelectorTerm in NodeSelector; for each single term, the MatchExpressions can not be nil or empty.

the documentation is pretty clear, and the implementation actually behaves as documented if it is given a NodeSelectorTerm that is empty (`v1helper.NodeSelectorRequirementsAsSelector(req.MatchExpressions)` returns `labels.Nothing()` if there are no match expressions)

I would pick a fix to validation in 1.10 to tolerate missing MatchExpressions:
```diff
 func ValidateNodeSelectorTerm(term core.NodeSelectorTerm, fldPath *field.Path) field.ErrorList {
        allErrs := field.ErrorList{}

-       if len(term.MatchExpressions) == 0 {
-               return append(allErrs, field.Required(fldPath.Child("matchExpressions"), "must have at least one node selector requirement"))
-       }
        for j, req := range term.MatchExpressions {
                allErrs = append(allErrs, ValidateNodeSelectorRequirement(req, fldPath.Child("matchExpressions").Index(j))...)
        }
```

cc @msau42 since this is called from validateStorageNodeAffinityAnnotation as well